sys/usb/usbus/hid/hid.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (C) 2020 Nils Ollrogge
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup usbus_hid
+ * @{
+ * @file
+ *
+ * @author  Nils Ollrogge <nils-ollrogge@outlook.de>
+ * @}
+ */
+
+#define USB_H_USER_IS_RIOT_INTERNAL
+
+#include <string.h>
+
+#include "usb/usbus.h"
+#include "usb/usbus/control.h"
+#include "usb/usbus/hid.h"
+#include "tsrb.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static void _init(usbus_t *usbus, usbus_handler_t *handler);
+static void _event_handler(usbus_t *usbus, usbus_handler_t *handler,
+                           usbus_event_usb_t event);
+static int _control_handler(usbus_t *usbus, usbus_handler_t *handler,
+                            usbus_control_request_state_t state,
+                            usb_setup_t *setup);
+static void _transfer_handler(usbus_t *usbus, usbus_handler_t *handler,
+                              usbdev_ep_t *ep, usbus_event_transfer_t event);
+
+static void _handle_tx_ready(event_t *ev);
+
+static const usbus_handler_driver_t hid_driver = {
+    .init = _init,
+    .event_handler = _event_handler,
+    .control_handler = _control_handler,
+    .transfer_handler = _transfer_handler
+};
+
+static size_t _gen_hid_descriptor(usbus_t *usbus, void *arg);
+
+static const usbus_descr_gen_funcs_t _hid_descriptor = {
+    .fmt_post_descriptor = _gen_hid_descriptor,
+    .len = {
+        .fixed_len = sizeof(usb_desc_hid_t)
+    },
+    .len_type = USBUS_DESCR_LEN_FIXED
+};
+
+static size_t _gen_hid_descriptor(usbus_t *usbus, void *arg)
+{
+    usbus_hid_device_t *hid_dev = arg;
+    usb_desc_hid_t hid_desc;
+
+    hid_desc.length = sizeof(usb_desc_hid_t);
+    hid_desc.desc_type = USB_HID_DESCR_HID;
+    hid_desc.bcd_hid = USB_HID_VERSION_BCD;
+    hid_desc.country_code = USB_HID_COUNTRY_CODE_NOTSUPPORTED;
+    hid_desc.num_descrs = 0x01;
+    hid_desc.report_type = USB_HID_DESCR_REPORT;
+    hid_desc.report_length = hid_dev->report_desc_size;
+
+    usbus_control_slicer_put_bytes(usbus, (uint8_t *)&hid_desc,
+                                   sizeof(hid_desc));
+    return sizeof(usb_desc_hid_t);
+}
+
+static void _handle_tx_ready(event_t *ev)
+{
+    usbus_hid_device_t *hid = container_of(ev, usbus_hid_device_t, tx_ready);
+
+    /* linked list ep at index 1 is IN */
+    usbdev_ep_ready(hid->iface.ep->next->ep, hid->occupied);
```

Comment:
I think it will be better and less error prone to have something like
`hid->ep_name `